### PR TITLE
8258077: Using -Xcheck:jni can lead to a double-free after JDK-8193234

### PR DESCRIPTION
**Pull request by dholmes-ora**
The fix in JDK-8193234 had an unintended consequence for the Release<X>ArrayElements API, which is now fixed in this issue. 

I'd like to thank Mauro Lacy and Dmitry Timofeev for raising, analysing and discussing this issue. You can follow the thread here:

http://mail.openjdk.java.net/pipermail/hotspot-dev/2020-December/047248.html

Although the fix itself is very basic I found a couple of other issues along the way, so I have staged the commits as follows for ease of understanding:

Step 1: added a regression test for the current double-free problem
Step 2: Only try to print the GuardedMemory info in debug builds as it can lead to secondary crashes
Step 3: Fix incorrect function names in the error messages and cleanup formatting
Step 4: Revert the change from JDK-8193234
Step 5: Add the memory-leak test from JDK-8193234 as a manual test
Step 6: Fix the JNI_COMMIT memory leak only for the case of ReleasePrimitiveArrayCritical

Finally I had to tweak the test to fix a nativepath problem. 

Testing:
 - tiers 1-4 (tier 4 includes the test run with -Xcheck:jni)
 - local testing of the new tests

Thanks,
David

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258077](https://bugs.openjdk.java.net/browse/JDK-8258077): Using -Xcheck:jni can lead to a double-free after JDK-8193234

### Reviewers
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to 8786d584b2c5d04885a12e89ecd7c5ebd4a88932
 * [Harold Seigel](https://openjdk.java.net/census#hseigel) (@hseigel - **Reviewer**) ⚠️ Review applies to b989bc52fef075565bb197ed35276e37999d00a5

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1816/head:pull/1816`
`$ git checkout pull/1816`

### CODE REVIEW

File: test/hotspot/jtreg/runtime/jni/checked/libTestCheckedReleaseCriticalArray.c [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  if (isCopy == JNI_FALSE) {
+    jint len = (*env)->GetArrayLength(env, iarr);
+    // make arbitrary changes to the array
+    for (int i = 0; i < len; i++) {
+      arr[i] *= 2;
+    }
+    // write-back using JNI_COMMIT to test for memory leak
+    (*env)->ReleasePrimitiveArrayCritical(env, iarr, arr, JNI_COMMIT);
+  }
+  // we skip the test is the VM makes a copy - as it will definitely leak
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/jni/checked/libTestCheckedReleaseCriticalArray.c:44]: typo - s/is the/if the/

**Review by dcubed-ojdk — APPROVED** [review]
I only have nits so it's your call on whether to make the changes.
I like the new tests! Thanks for doing that. In what Tier do these
new tests execute? Tier4 with "-Xcheck:jni" is a given, but do
they also run in an earlier Tier or two?

**Review by hseigel — APPROVED** [review]
These changes look good!  Are any changes needed to the JNI Spec?
Thanks, Harold

**Pull request closed without merging**